QKSMS/src/main/java/com/mariussoft/endlessjabber/sdk/EndlessJabberWakefulService.java

```diff
@@ -7,72 +7,73 @@
-			}
-			}
-		} catch (InstantiationException | IllegalAccessException | ClassNotFoundException e) {
-
-		}
-
-		EndlessJabberReceiver.completeWakefulIntent(intent);
-	}
+    public EndlessJabberWakefulService() {
+        super("EndlessJabberWakefulService");
+    }
+
+    @Override
+    protected void onHandleIntent(Intent intent) {
+        Context context = getApplicationContext();
+
+        // Get the class to call
+        String implementationClass = context.getSharedPreferences("EndlessJabberSDK", Context.MODE_PRIVATE).getString("InterfaceClass", null);
+
+        if (implementationClass != null) {
+            IEndlessJabberImplementation instanceOfMyClass;
+            try {
+                instanceOfMyClass = (IEndlessJabberImplementation) Class.forName(implementationClass).newInstance();
+
+                Bundle extras = intent.getExtras();
+
+                switch (extras.getString("Action")) {
+                    case "UpdateRead": {
+                        long time = extras.getLong("Time");
+                        int conversationID = extras.getInt("ConversationID");
+                        instanceOfMyClass.UpdateReadMessages(context, time, conversationID);
+                        break;
+                    }
+                    case "DeleteThread": {
+
+                        int conversationID = extras.getInt("ConversationID");
+                        instanceOfMyClass.DeleteThread(context, conversationID);
+                        break;
+                    }
+                    case "SendMMS": {
+
+                        boolean save = extras.getBoolean("Save");
+                        boolean send = extras.getBoolean("Send");
+                        String[] Recipients = extras.getStringArray("Recipients");
+
+                        MMSPart[] parts = new MMSPart[extras.getParcelableArray("Parts").length];
+
+                        for (int i = 0; i < parts.length; i++) {
+                            parts[i] = (MMSPart) extras.getParcelableArray("Parts")[i];
+                        }
+
+                        instanceOfMyClass.SendMMS(context, Recipients, parts, extras.getString("Subject"), save, send);
+                        break;
+                    }
+                    case "SendSMS": {
+                        boolean send = extras.getBoolean("Send");
+                        String[] Recipients = extras.getStringArray("Recipients");
+                        String message = extras.getString("Message");
+
+                        instanceOfMyClass.SendSMS(context, Recipients, message, send);
+                        break;
+                    }
+                    case "UpdateInfo": {
+                        EndlessJabberInterface.SendInfoToEndlessJabber(context);
+                        break;
+                    }
+                }
+            } catch (InstantiationException e) {
+
+            } catch (IllegalAccessException e) {
+
+            } catch (ClassNotFoundException e) {
```

Comment:
Try-with-resources is the one you're thinking of I reckon, Android Studio doesn't like the multicatch though andis drawing all sorts of red squiggles. The [release notes for 0.3.2](http://tools.android.com/recent/androidstudio032released) suggest that it's fine to use so if you're confident I'll undo that change tomorrow when I'm at my laptop (and test on a jellybean device)